### DISCUSSION TITLE
A collection of LuaSnip guides

### DISCUSSION THREAD

**Discussion by evesdropper** [L3MON4D3/LuaSnip#762] [Show and tell]
I originally made a guide for moving from UltiSnips to LuaSnip, but in recent times I've made a few more guides:

- Original LuaSnip Guide - "From UltiSnips to LuaSnip: A Comprehensive Guide": [[md]](https://github.com/evesdropper/dotfiles/tree/main/nvim/lua/snippets) [[website]](https://evesdropper.dev/files/luasnip/2022/11/25/luasnip-guide.html)
- Condition Objects Guide - "LuaSnip Condition Objects: Use Python snippets within LaTeX Code Listings": [[md]](https://github.com/evesdropper/dotfiles/blob/main/nvim/lua/snippets/condition-objects.md) [[website]](https://evesdropper.dev/files/luasnip/2022/12/31/luasnip-minted.html)
- Choice/Dynamic Nodes Guide - "Advanced LuaSnip: Modular Snippets with Choice/Dynamic Nodes": [[md]](https://github.com/evesdropper/dotfiles/blob/main/nvim/lua/snippets/choice-dynamic-nodes.md) [[website]](https://evesdropper.dev/files/luasnip/2023/02/05/dynamic-choice-luasnip.html)

A majority of them are in LaTeX context but most of the concepts can be applied to other languages as well. Feedback and ideas are greatly appreciated!

Also yes some of the website is scuffed :sob: I'll probably fix that when I do the next big website fix.

**Comment by L3MON4D3** [comment]
Ohh awesome!
It's really nice to see how useful luasnip is for you :D

Few points:
* Nice use of condition objects! I really like the boolean algebra, but it's not really applicable often. If it is, it looks cool xD
For your usecase, "injecting" snippets of one filetype into another, [`ft_func`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#config-reference) (and the implementation of the [defaults](https://github.com/L3MON4D3/LuaSnip/blob/4eead79ef0f93a3786ca5b6265294ea94849586d/lua/luasnip/extras/filetype_functions.lua)) might also be interesting: you could enable all snippets registered for python by adding `"python"` to the filetypes returned from `ft_func` in the minted-python environment (of course, then this selectivity from setting the condition on certain snippets only is gone)
* Obviously, also always happy to see restoreNode :D
I'd give them more descriptive keys than `""`, that might make clear that they are linked together, and that multiple restoreNodes can coexist in one snippet.
* Using choiceNodes to group related snippets together isn't used very often (at least I don't see it often :D), but it's really useful to prevent overly descriptive triggers :+1:
[Here's](https://github.com/L3MON4D3/Dotfiles/blob/362701be7711049196a92bc12d40a8a38b7dba03/.config/nvim/luasnippets/tex.lua#L122-L154) a set of snippets I used this in
* I think the label-generation-snippets don't even need a dynamicNode, snippets can just contain a snippetNode like they can contain an insertNode. Replacing `d(generate_label, {}, {user_args = {...}})` with just `generate_label(...)` (and adjusting `generate_label`s parameters) might work without further changes

Overall, really nice guides/showcases, I like the focus on useful examples :ok_hand:
Would you do a PR to add them in the readme?

**Reply by evesdropper** [reply]
Thanks for the feedback!

> Nice use of condition objects! I really like the boolean algebra, but it's not really applicable often. If it is, it looks cool xD

Probably not incredibly applicable or immediately usable for every case, but I could definitely see cases where you could combine some basic conditions together rather than making a very large/overly specific function (e.g. I think I used `line_begin` and `itemize`/`enumerate` environment for a snippet in my LaTeX snippets rather than making that a double function). Perhaps that's another point for modularity.

> For your usecase, "injecting" snippets of one filetype into another, ft_func (and the implementation of the defaults) might also be interesting: you could enable all snippets registered for python by adding "python" to the filetypes returned from ft_func in the minted-python environment (of course, then this selectivity from setting the condition on certain snippets only is gone)

That sounds like a great option too; how would one go about implementing that?

> I'd give them more descriptive keys than "", that might make clear that they are linked together, and that multiple restoreNodes can coexist in one snippet.

Sounds good - yeah it might be good to put something other than an empty string :rofl:, both for the reader/user and me when authoring snippets. 

> I think the label-generation-snippets don't even need a dynamicNode, snippets can just contain a snippetNode like they can contain an insertNode. Replacing d(generate_label, {}, {user_args = {...}}) with just generate_label(...) (and adjusting generate_labels parameters) might work without further changes

Hmm, so something like changing the function signature to `generate_label(args, kwargs)` and then just directly calling it should be fine?

~~Note: this may well be another way to "break" LuaSnip - finding a function to save a lot of snippet writing :eyes:~~ and oops i procrastinated my math homework again for [this](https://github.com/evesdropper/dotfiles/blob/main/nvim/luasnip/tex.lua#L220-L242) :rofl: 

> Overall, really nice guides/showcases, I like the focus on useful examples ok_hand
Would you do a PR to add them in the readme?

Would be happy to! I'm probably going to make a few small changes to the website presentation to support future guides as well as implement some of the ideas you suggested before doing so of course :smile:

**Comment by L3MON4D3** [comment]
> Probably not incredibly applicable or immediately usable for every case, but I could definitely see cases where you could combine some basic conditions together rather than making a very large/overly specific function

Yeah true. I haven't had any chance to do so yet though :D

> That sounds like a great option too; how would one go about implementing that?

Check the `from_cursor_or_filetype`-function in the file I linked before, it resolves the treesitter-languages at the cursor

> and oops i procrastinated my math homework again for this 🤣

Yes like that, nice!! Finally the "decision" (very conscious ofc xD) to have snippets authored in lua is paying off :D

**Reply by evesdropper** [reply]
> Check the from_cursor_or_filetype-function in the file I linked before, it resolves the treesitter-languages at the cursor

:eyes: I think I have an idea on how to implement this. Maybe it will be more efficient instead of using a ton of conditions.

> Yes like that, nice!! Finally the "decision" (very conscious ofc xD) to have snippets authored in lua is paying off :D

Yep for sure - there's a lot I have in mind but it'll probably be a while until I finish implementing it all. It's so messy right now :rofl: 

:despair: Sometimes I wish I could share the beauty of LuaSnip to others (my friends prefer to use vscode so it's not like I can teach anyone)

**Reply by L3MON4D3** [reply]
> :eyes: I think I have an idea on how to implement this

:ok_hand::ok_hand:

> Sometimes I wish I could share the beauty of LuaSnip to others (my friends prefer to use vscode so it's not like I can teach anyone)

Yeah, feels like neovim users are far and between in the real world xD
But your blogposts are doing a nice job at teaching/sharing about luasnip, at least to other online people :D

**Comment by atticus-sullivan** [comment]
Nice guides and showcases. Always cool to get new ideas for how to use luaSnip (and inspiration for LaTeX snippets). Just one note your github links are broken (the ones to your website are fine though).

**Reply by evesdropper** [reply]
Thanks - I've got into a habit of wasting time working on some of the most silly snippets ever :laughing: (thanks live lectures). I think I should've fixed the md links now.

**Reply by atticus-sullivan** [reply]
Yes, works now :+1:
